#include "useritem.h"
#include "ui_useritem.h"
#include <QDebug>
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QDate>

UserItem::UserItem(QWidget *parent) :
    QWidget(parent),
    ui(new Ui::UserItem),
    currentStatus("æ­£å¸¸")
{
    ui->setupUi(this);

    qDebug() << "=== å¼€å§‹åˆ›å»ºUserItem ===";

    // ============================================
    // å…³é”®æ­¥éª¤ï¼šæ¸…é™¤å›ºå®šgeometryï¼Œä½¿ç”¨å¸ƒå±€ç®¡ç†å™¨
    // ============================================

    // æ¸…é™¤æ‰€æœ‰æ§ä»¶çš„å›ºå®šgeometry
    ui->pushButton->setGeometry(QRect());
    ui->pushButton_2->setGeometry(QRect());
    ui->pushButton_3->setGeometry(QRect());
    ui->label_3->setGeometry(QRect());
    ui->userName->setGeometry(QRect());
    ui->label->setGeometry(QRect());

    // æ¸…é™¤å¯èƒ½å­˜åœ¨çš„æ—§å¸ƒå±€
    if (this->layout()) {
        qDebug() << "åˆ é™¤UserItemæ—§å¸ƒå±€";
        delete this->layout();
    }

    // è®¾ç½®å¤§å°ç­–ç•¥
    setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
    setMinimumHeight(150);
    setMaximumHeight(180);

    // ============================================
    // åˆ›å»ºæ–°çš„å¸ƒå±€
    // ============================================

    // ä¸»æ°´å¹³å¸ƒå±€
    QHBoxLayout *mainLayout = new QHBoxLayout(this);
    mainLayout->setSpacing(20);
    mainLayout->setContentsMargins(25, 20, 25, 20);

    // å·¦ä¾§ï¼šå¤´åƒåŒºåŸŸ
    QVBoxLayout *avatarLayout = new QVBoxLayout();
    avatarLayout->setAlignment(Qt::AlignCenter);

    if (ui->label_3) {
        ui->label_3->setFixedSize(60, 60);
        ui->label_3->setStyleSheet(R"(
            QLabel {
                border-radius: 30px;
                background-color: #ecf0f1;
                border: 2px solid #bdc3c7;
                qproperty-alignment: AlignCenter;
            }
        )");
        avatarLayout->addWidget(ui->label_3);
    }

    mainLayout->addLayout(avatarLayout);

    // ä¸­é—´ï¼šç”¨æˆ·ä¿¡æ¯åŒºåŸŸï¼ˆå‚ç›´å¸ƒå±€ï¼‰
    QVBoxLayout *infoLayout = new QVBoxLayout();
    infoLayout->setSpacing(10);
    infoLayout->setContentsMargins(0, 5, 0, 5);

    // ç”¨æˆ·ä¿¡æ¯è¡Œï¼ˆæ°´å¹³å¸ƒå±€ï¼‰
    QHBoxLayout *userInfoRow = new QHBoxLayout();
    userInfoRow->setSpacing(30);

    // ç”¨æˆ·å
    if (ui->userName) {
        ui->userName->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
        ui->userName->setStyleSheet(R"(
            QLabel {
                font-size: 16px;
                font-weight: bold;
                color: #2c3e50;
                min-width: 120px;
            }
        )");
        userInfoRow->addWidget(ui->userName);
    }

    // æ³¨å†Œæ—¥æœŸ
    if (ui->label) {
        ui->label->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Fixed);
        ui->label->setStyleSheet(R"(
            QLabel {
                font-size: 14px;
                color: #7f8c8d;
                min-width: 150px;
            }
        )");
        userInfoRow->addWidget(ui->label);
    }

    userInfoRow->addStretch();
    infoLayout->addLayout(userInfoRow);

    // çŠ¶æ€æ˜¾ç¤ºåŒºåŸŸï¼ˆæ°´å¹³å¸ƒå±€ï¼‰
    QHBoxLayout *statusLayout = new QHBoxLayout();
    statusLayout->setSpacing(15);

    // çŠ¶æ€æ ‡ç­¾ï¼ˆåŠ¨æ€æ·»åŠ ï¼‰
    QLabel *statusLabel = new QLabel();
    statusLabel->setFixedSize(80, 28);
    statusLabel->setStyleSheet(R"(
        QLabel {
            font-size: 12px;
            font-weight: bold;
            border-radius: 14px;
            padding: 2px 10px;
            qproperty-alignment: AlignCenter;
            background-color: #27ae60;
            color: white;
        }
    )");
    statusLabel->setText("æ­£å¸¸");
    statusLayout->addWidget(statusLabel);

    // ä¸¾æŠ¥æ¬¡æ•°æ ‡ç­¾ï¼ˆåŠ¨æ€æ·»åŠ ï¼‰
    QLabel *reportLabel = new QLabel("è¢«ä¸¾æŠ¥: 0æ¬¡");
    reportLabel->setFixedSize(100, 28);
    reportLabel->setStyleSheet(R"(
        QLabel {
            font-size: 12px;
            color: #e74c3c;
            font-weight: bold;
            background-color: rgba(231, 76, 60, 0.1);
            border-radius: 14px;
            padding: 2px 10px;
            qproperty-alignment: AlignCenter;
        }
    )");
    statusLayout->addWidget(reportLabel);

    statusLayout->addStretch();
    infoLayout->addLayout(statusLayout);

    mainLayout->addLayout(infoLayout);

    // å³ä¾§ï¼šæ“ä½œæŒ‰é’®åŒºåŸŸï¼ˆå‚ç›´å¸ƒå±€ï¼‰
    QVBoxLayout *buttonLayout = new QVBoxLayout();
    buttonLayout->setSpacing(10);
    buttonLayout->setAlignment(Qt::AlignCenter);

    // ç¬¬ä¸€è¡ŒæŒ‰é’®ï¼ˆè§£å°å’Œè­¦å‘Šï¼‰
    QHBoxLayout *buttonRow1 = new QHBoxLayout();
    buttonRow1->setSpacing(15);

    if (ui->pushButton) {
        ui->pushButton->setFixedSize(100, 36);
        ui->pushButton->setStyleSheet(R"(
            QPushButton {
                background-color: #27ae60;
                color: white;
                border: none;
                border-radius: 8px;
                font-size: 14px;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #229954;
            }
            QPushButton:disabled {
                background-color: #bdc3c7;
            }
        )");
        buttonRow1->addWidget(ui->pushButton);

        connect(ui->pushButton, &QPushButton::clicked, this, [this]() {
            qDebug() << "ğŸ”“ è§£å°ç”¨æˆ·:" << (ui->userName ? ui->userName->text() : "æœªçŸ¥ç”¨æˆ·");
            currentStatus = "æ­£å¸¸";
            updateButtonsByStatus();
            emit userUnbanned();
        });
    }

    if (ui->pushButton_3) {
        ui->pushButton_3->setFixedSize(100, 36);
        ui->pushButton_3->setStyleSheet(R"(
            QPushButton {
                background-color: #f39c12;
                color: white;
                border: none;
                border-radius: 8px;
                font-size: 14px;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #d68910;
            }
            QPushButton:disabled {
                background-color: #bdc3c7;
            }
        )");
        buttonRow1->addWidget(ui->pushButton_3);

        connect(ui->pushButton_3, &QPushButton::clicked, this, [this]() {
            qDebug() << "âš ï¸ è­¦å‘Šç”¨æˆ·:" << (ui->userName ? ui->userName->text() : "æœªçŸ¥ç”¨æˆ·");
            currentStatus = "è­¦å‘Šä¸­";
            updateButtonsByStatus();
            emit userWarned();
        });
    }

    buttonLayout->addLayout(buttonRow1);

    // ç¬¬äºŒè¡ŒæŒ‰é’®ï¼ˆå°ç¦ï¼‰
    QHBoxLayout *buttonRow2 = new QHBoxLayout();
    buttonRow2->setSpacing(15);

    if (ui->pushButton_2) {
        ui->pushButton_2->setFixedSize(215, 36); // å®½åº¦ç­‰äºä¸¤ä¸ªæŒ‰é’®å®½åº¦åŠ é—´è·
        ui->pushButton_2->setStyleSheet(R"(
            QPushButton {
                background-color: #e74c3c;
                color: white;
                border: none;
                border-radius: 8px;
                font-size: 14px;
                font-weight: bold;
            }
            QPushButton:hover {
                background-color: #c0392b;
            }
            QPushButton:disabled {
                background-color: #bdc3c7;
            }
        )");
        buttonRow2->addWidget(ui->pushButton_2);

        connect(ui->pushButton_2, &QPushButton::clicked, this, [this]() {
            qDebug() << "ğŸš« å°ç¦ç”¨æˆ·:" << (ui->userName ? ui->userName->text() : "æœªçŸ¥ç”¨æˆ·");
            currentStatus = "å·²å°ç¦";
            updateButtonsByStatus();
            emit userBanned();
        });
    }

    buttonLayout->addLayout(buttonRow2);
    mainLayout->addLayout(buttonLayout);

    // è®¾ç½®UserItemæ•´ä½“æ ·å¼
    setStyleSheet(R"(
        UserItem {
            background-color: white;
            border: 1px solid #dfe6e9;
            border-radius: 12px;
            margin: 8px 0;
        }
        UserItem:hover {
            border-color: #3498db;
            box-shadow: 0 2px 12px rgba(52, 152, 219, 0.15);
        }
    )");

    setAutoFillBackground(true);

    // å­˜å‚¨åŠ¨æ€åˆ›å»ºçš„æ§ä»¶æŒ‡é’ˆï¼Œä»¥ä¾¿åç»­æ›´æ–°
    // è¿™é‡Œå¯ä»¥ä½¿ç”¨QObjectçš„setPropertyæ¥å­˜å‚¨é¢å¤–ä¿¡æ¯
    setProperty("statusLabel", QVariant::fromValue<QWidget*>(statusLabel));
    setProperty("reportLabel", QVariant::fromValue<QWidget*>(reportLabel));

    qDebug() << "=== UserItemåˆ›å»ºå®Œæˆ ===";
}

UserItem::~UserItem()
{
    delete ui;
}

void UserItem::setUserData(const QString &userId,
                          const QString &username,
                          const QString &registerDate,
                          const QString &status,
                          int reportCount,
                          const QString &email)
{
    this->userId = userId;
    this->currentStatus = status;

    qDebug() << "è®¾ç½®ç”¨æˆ·æ•°æ® - ID:" << userId << "ç”¨æˆ·å:" << username;

    // è®¾ç½®ç”¨æˆ·å
    if (ui->userName) {
        ui->userName->setText(username);
    }

    // è®¾ç½®æ³¨å†Œæ—¥æœŸ
    if (ui->label) {
        ui->label->setText("æ³¨å†Œ: " + registerDate);
    }

    // æ›´æ–°çŠ¶æ€æ˜¾ç¤º
    updateButtonsByStatus();

    // æ›´æ–°çŠ¶æ€æ ‡ç­¾
    QWidget* statusLabelWidget = property("statusLabel").value<QWidget*>();
    if (statusLabelWidget) {
        QLabel* statusLabel = qobject_cast<QLabel*>(statusLabelWidget);
        if (statusLabel) {
            QString statusText = status;
            QString statusStyle;

            if (status == "æ­£å¸¸") {
                statusStyle = "background-color: #27ae60; color: white;";
            } else if (status == "è­¦å‘Šä¸­") {
                statusStyle = "background-color: #f39c12; color: white;";
            } else if (status == "å·²å°ç¦") {
                statusStyle = "background-color: #e74c3c; color: white;";
            } else {
                statusStyle = "background-color: #95a5a6; color: white;";
            }

            statusLabel->setText(statusText);
            statusLabel->setStyleSheet(QString(R"(
                QLabel {
                    font-size: 12px;
                    font-weight: bold;
                    border-radius: 14px;
                    padding: 2px 10px;
                    qproperty-alignment: AlignCenter;
                    %1
                }
            )").arg(statusStyle));
        }
    }

    // æ›´æ–°ä¸¾æŠ¥æ¬¡æ•°æ ‡ç­¾
    QWidget* reportLabelWidget = property("reportLabel").value<QWidget*>();
    if (reportLabelWidget) {
        QLabel* reportLabel = qobject_cast<QLabel*>(reportLabelWidget);
        if (reportLabel) {
            reportLabel->setText(QString("è¢«ä¸¾æŠ¥: %1æ¬¡").arg(reportCount));

            // æ ¹æ®ä¸¾æŠ¥æ¬¡æ•°è®¾ç½®é¢œè‰²
            QString reportStyle;
            if (reportCount == 0) {
                reportStyle = "color: #27ae60; background-color: rgba(39, 174, 96, 0.1);";
            } else if (reportCount < 3) {
                reportStyle = "color: #f39c12; background-color: rgba(243, 156, 18, 0.1);";
            } else {
                reportStyle = "color: #e74c3c; background-color: rgba(231, 76, 60, 0.1);";
            }

            reportLabel->setStyleSheet(QString(R"(
                QLabel {
                    font-size: 12px;
                    font-weight: bold;
                    border-radius: 14px;
                    padding: 2px 10px;
                    qproperty-alignment: AlignCenter;
                    %1
                }
            )").arg(reportStyle));
        }
    }

    // è®¾ç½®å¤´åƒï¼ˆæ ¹æ®ç”¨æˆ·åé¦–å­—æ¯ï¼‰
    if (ui->label_3) {
        if (!username.isEmpty()) {
            QString firstLetter = username.left(1).toUpper();

            // æ ¹æ®é¦–å­—æ¯è®¾ç½®ä¸åŒçš„èƒŒæ™¯è‰²
            QChar firstChar = firstLetter.at(0);
            QString bgColor;

            if (firstChar >= 'A' && firstChar <= 'F') {
                bgColor = "#3498db";  // è“è‰²
            } else if (firstChar >= 'G' && firstChar <= 'L') {
                bgColor = "#9b59b6";  // ç´«è‰²
            } else if (firstChar >= 'M' && firstChar <= 'R') {
                bgColor = "#2ecc71";  // ç»¿è‰²
            } else {
                bgColor = "#e67e22";  // æ©™è‰²
            }

            ui->label_3->setStyleSheet(QString(R"(
                QLabel {
                    border-radius: 30px;
                    background-color: %1;
                    border: 2px solid %2;
                    color: white;
                    font-size: 24px;
                    font-weight: bold;
                    qproperty-alignment: AlignCenter;
                }
            )").arg(bgColor).arg(QColor(bgColor).darker(120).name()));

            ui->label_3->setText(firstLetter);
        }
    }

    qDebug() << "ç”¨æˆ·æ•°æ®è®¾ç½®å®Œæˆ - çŠ¶æ€:" << status << "ä¸¾æŠ¥æ¬¡æ•°:" << reportCount;
}

void UserItem::updateButtonsByStatus()
{
    qDebug() << "æ›´æ–°æŒ‰é’®çŠ¶æ€ - å½“å‰çŠ¶æ€:" << currentStatus;

    // æ ¹æ®å½“å‰çŠ¶æ€å¯ç”¨/ç¦ç”¨æŒ‰é’®
    if (ui->pushButton) {    // è§£å°æŒ‰é’®
        ui->pushButton->setEnabled(currentStatus == "å·²å°ç¦");
        if (currentStatus == "å·²å°ç¦") {
            ui->pushButton->setStyleSheet(R"(
                QPushButton {
                    background-color: #27ae60;
                    color: white;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
                QPushButton:hover {
                    background-color: #229954;
                }
            )");
        } else {
            ui->pushButton->setStyleSheet(R"(
                QPushButton {
                    background-color: #bdc3c7;
                    color: #7f8c8d;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
            )");
        }
    }

    if (ui->pushButton_2) {  // å°ç¦æŒ‰é’®
        ui->pushButton_2->setEnabled(currentStatus != "å·²å°ç¦");
        if (currentStatus != "å·²å°ç¦") {
            ui->pushButton_2->setStyleSheet(R"(
                QPushButton {
                    background-color: #e74c3c;
                    color: white;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
                QPushButton:hover {
                    background-color: #c0392b;
                }
            )");
        } else {
            ui->pushButton_2->setStyleSheet(R"(
                QPushButton {
                    background-color: #bdc3c7;
                    color: #7f8c8d;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
            )");
        }
    }

    if (ui->pushButton_3) {  // è­¦å‘ŠæŒ‰é’®
        ui->pushButton_3->setEnabled(currentStatus != "å·²å°ç¦");
        if (currentStatus != "å·²å°ç¦") {
            ui->pushButton_3->setStyleSheet(R"(
                QPushButton {
                    background-color: #f39c12;
                    color: white;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
                QPushButton:hover {
                    background-color: #d68910;
                }
            )");
        } else {
            ui->pushButton_3->setStyleSheet(R"(
                QPushButton {
                    background-color: #bdc3c7;
                    color: #7f8c8d;
                    border: none;
                    border-radius: 8px;
                    font-size: 14px;
                    font-weight: bold;
                }
            )");
        }
    }

    // æ›´æ–°æŒ‰é’®æ–‡æœ¬
    if (currentStatus == "è­¦å‘Šä¸­") {
        if (ui->pushButton_3) {
            ui->pushButton_3->setText("æ’¤é”€è­¦å‘Š");
        }
    } else {
        if (ui->pushButton_3) {
            ui->pushButton_3->setText("è­¦å‘Š");
        }
    }
}
